File: csharp/behavioral/rare_interpreter.cs
namespace Gof.ModernCSharp.Behavioral;

// Interpreter models a tiny language as expression objects that evaluate themselves.
// Modern C# note: rare unless building a DSL. Parser libraries, expression trees,
// pattern matching, or simple functions are often better.
public static class RareInterpreter
{
    interface IExpression { int Interpret(); }
    sealed record Number(int Value) : IExpression { public int Interpret() => Value; }
    sealed record Add(IExpression Left, IExpression Right) : IExpression
    {
        public int Interpret() => Left.Interpret() + Right.Interpret();
    }

    public static string Run() => new Add(new Number(2), new Add(new Number(3), new Number(4))).Interpret().ToString();
}

File: py/behavioral/rare_interpreter.py
# Interpreter models a tiny language as expression objects that evaluate themselves.
# Modern Python note: this is uncommon unless you are building a DSL. Parser
# libraries, pattern matching, or plain functions are often simpler.
class Number:
    def __init__(self, value):
        self.value = value

    def interpret(self):
        return self.value


class Add:
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self):
        return self.left.interpret() + self.right.interpret()


if __name__ == "__main__":
    expression = Add(Number(2), Add(Number(3), Number(4)))
    print(expression.interpret())
